src/hotspot/share/prims/jvmtiEnvBase.cpp

```diff
@@ -534,29 +536,34 @@ void JvmtiEnvBase::destroy_jni_reference(JavaThread *thread, jobject jobj) {
 // Threads
 //
 
-jobject *
-JvmtiEnvBase::new_jobjectArray(int length, Handle *handles) {
+jthread *
+JvmtiEnvBase::new_jthreadArray(int length, Handle *handles) {
   if (length == 0) {
     return NULL;
   }
 
-  jobject *objArray = (jobject *) jvmtiMalloc(sizeof(jobject) * length);
+  jthread* objArray = (jthread *) jvmtiMalloc(sizeof(jthread) * length);
   NULL_CHECK(objArray, NULL);
 
-  for (int i=0; i<length; i++) {
-    objArray[i] = jni_reference(handles[i]);
+  for (int i = 0; i < length; i++) {
+    objArray[i] = (jthread)jni_reference(handles[i]);
   }
   return objArray;
 }
 
-jthread *
-JvmtiEnvBase::new_jthreadArray(int length, Handle *handles) {
-  return (jthread *) new_jobjectArray(length,handles);
-}
-
 jthreadGroup *
-JvmtiEnvBase::new_jthreadGroupArray(int length, Handle *handles) {
-  return (jthreadGroup *) new_jobjectArray(length,handles);
+JvmtiEnvBase::new_jthreadGroupArray(int length, objArrayHandle groups) {
+  if (length == 0) {
+    return NULL;
+  }
+
+  jthreadGroup* objArray = (jthreadGroup *) jvmtiMalloc(sizeof(jthreadGroup) * length);
+  NULL_CHECK(objArray, NULL);
+
+  for (int i = 0; i < length; i++) {
+    objArray[i] = (jthreadGroup)JNIHandles::make_local(groups->obj_at(i));
```

Comment:
Nit: It is better to use `jni_reference` instead of `JNIHandles::make_local` for consistency as at the line 549.

Reply:
jni_reference takes a Handle and this passes a Handle to an objArray so groups->obj_at(i) is an oop and it's a waste to make it a handle for jni_reference to just unhandlize it.